### PR TITLE
maybeReturnSc_corres

### PR DESCRIPTION
**Pull request by MitchellBuckley**

### CODE REVIEW

File: proof/refine/ARM/Ipc_R.thy [hunk 1]
```diff
@@ -2990,6 +2990,163 @@ lemma do_nbrecv_failed_transfer_corres:
...
+  apply (wpsimp wp: get_object_wp)
+  done
+
+lemma set_sc_obj_ref_valid_tcbs[wp]:
+  "set_sc_obj_ref f scp v \<lbrace>valid_tcbs\<rbrace>"
+  unfolding valid_tcbs_def
+  apply (wpsimp wp: hoare_vcg_all_lift hoare_vcg_imp_lift')
+  done
+
+lemma set_sc_obj_ref_ko_not_sftcb_at[wp]:
```
Comment (MitchellBuckley) [inline proof/refine/ARM/Ipc_R.thy:3015]: rename

File: proof/refine/ARM/Ipc_R.thy [hunk 2]
```diff
@@ -2990,6 +2990,163 @@ lemma do_nbrecv_failed_transfer_corres:
...
+  "\<lbrace>\<lambda>s. \<not> ko_at (TCB tcb) t s \<and> (t = tptr \<longrightarrow> obj_at (\<lambda>ko. \<exists>tcb'. ko = TCB tcb' \<and> f tcb' \<noteq> tcb) tptr s)\<rbrace>
+  thread_set f tptr
+  \<lbrace>\<lambda>_ s. \<not> ko_at (TCB tcb) t s\<rbrace>"
+  apply (wpsimp simp: thread_set_def set_object_def obj_at_def pred_neg_def
+                  wp: get_object_wp)+
+  apply (clarsimp simp: obj_at_def dest!: get_tcb_SomeD)
+  done
+
+lemma set_sc_obj_resff_valid_tcb[wp]:
```
Comment (MitchellBuckley) [inline proof/refine/ARM/Ipc_R.thy:3024]: rename

File: proof/refine/ARM/Ipc_R.thy [hunk 3]
```diff
@@ -2990,6 +2990,163 @@ lemma do_nbrecv_failed_transfer_corres:
...
+  "valid_tcb' tcb s' \<Longrightarrow> valid_tcb' (tcbSchedContext_update Map.empty tcb) s'"
+  by (auto simp: valid_tcb'_def valid_cap'_def tcb_cte_cases_def)
+
+lemma maybeReturnSc_corres:
+  "corres dc
+   (ntfn_at ntfnPtr and tcb_at thread and valid_tcbs and pspace_aligned
+      and scheduler_act_not thread
+      and pspace_distinct and weak_valid_sched_action
+      and not_queued thread and not_in_release_q thread)
+   (valid_tcbs' and valid_queues and valid_queues' and valid_release_queue and valid_release_queue')
```
Comment (mktnk3) [inline proof/refine/ARM/Ipc_R.thy:3093]: maybe use `valid_release_queue_iff`? We probably could do with the same for `valid_queues` and the other...

File: spec/haskell/src/SEL4/Object/SchedContext.lhs [hunk 4]
```diff
@@ -685,7 +685,7 @@ This module uses the C preprocessor to select a target architecture.
 > maybeReturnSc ntfnPtr tcbPtr = do
 >     nscOpt <- liftM ntfnSc $ getNotification ntfnPtr
 >     tscOpt <- threadGet tcbSchedContext tcbPtr
->     when (nscOpt == tscOpt) $ do
+>     when (nscOpt == tscOpt && nscOpt /= Nothing) $ do
 >         threadSet (\tcb -> tcb { tcbSchedContext = Nothing }) tcbPtr
 >         scPtr <- return $ fromJust nscOpt
 >         sc <- getSchedContext scPtr
```
Comment (mktnk3) [inline spec/haskell/src/SEL4/Object/SchedContext.lhs:691]: It looks like the Haskell and abstract versions of this function are different from the C. The C code doesn’t seem to do rescheduling. Do you mind checking this out, Mitch?

Comment (MitchellBuckley) [reply spec/haskell/src/SEL4/Object/SchedContext.lhs:691]: Will do.

File: proof/refine/ARM/Ipc_R.thy [hunk 5]
```diff
@@ -2990,6 +2990,163 @@ lemma do_nbrecv_failed_transfer_corres:
   unfolding do_nbrecv_failed_transfer_def doNBRecvFailedTransfer_def
   by (simp add: badgeRegister_def badge_register_def, rule user_setreg_corres)
 
+lemma ntfn_relation_par_inj:
```
Comment (MitchellBuckley) [inline proof/refine/ARM/Ipc_R.thy:2993]: move these up to the move file?

File: proof/refine/ARM/Ipc_R.thy [hunk 6]
```diff
@@ -2990,6 +2990,163 @@ lemma do_nbrecv_failed_transfer_corres:
...
+  apply (rename_tac tcb)
+  apply (prop_tac "valid_tcb t tcb s")
+   apply (clarsimp simp: valid_tcbs_def)
+   apply (drule_tac x=t in spec)
+   apply (drule_tac x=tcb in spec)
+   apply (clarsimp simp: obj_at_def)
+   apply (clarsimp simp: )
+  done
+
+lemma cross_rel_srE:
```
Comment (MitchellBuckley) [inline proof/refine/ARM/Ipc_R.thy:3058]: Should this have a better name? It should also be moved.

File: proof/refine/ARM/Ipc_R.thy [hunk 7]
```diff
@@ -2990,6 +2990,163 @@ lemma do_nbrecv_failed_transfer_corres:
...
+lemma thread_set_weak_valid_sched_action2:
+  "\<lbrace>weak_valid_sched_action and scheduler_act_not tptr\<rbrace> thread_set f tptr \<lbrace>\<lambda>rv. weak_valid_sched_action\<rbrace>"
+  apply (wpsimp wp: thread_set_wp simp: obj_at_kh_kheap_simps vs_all_heap_simps fun_upd_def
+                    weak_valid_sched_action_def)
+  apply (auto simp: weak_valid_sched_action_def scheduler_act_not_def)
+  apply (rule_tac x=ref' in exI; simp)
+  apply clarsimp
+ done
+
+definition notQueued :: "obj_ref \<Rightarrow> kernel_state \<Rightarrow> bool" where
```
Comment (MitchellBuckley) [inline proof/refine/ARM/Ipc_R.thy:3071]: Move this definition.

File: spec/haskell/src/SEL4/Object/SchedContext.lhs (hunk 4 above)
Comment (MitchellBuckley) [reply spec/haskell/src/SEL4/Object/SchedContext.lhs:691]: I hope that our email discussions have resolved this (This change is necessary but was lost in some changes on the systems side. It is now recorded in commits against cmillar/mcs-verified or cmillar/mcs-target)

Comment (xurtis) [reply spec/haskell/src/SEL4/Object/SchedContext.lhs:691]: It's currently tied up in `sporadic-fixes`, I should be able to clean that branch up and make some actual distinct PRs there next week

File: proof/refine/ARM/Ipc_R.thy [hunk 8]
```diff
@@ -2990,6 +2990,163 @@ lemma do_nbrecv_failed_transfer_corres:
...
+        apply (clarsimp simp: tcb_relation_def)
+       apply (wpsimp wp: thread_get_wp threadGet_wp)+
+      apply (frule ntfn_relation_par_inj, simp)
+     apply (wpsimp wp: get_simple_ko_wp getNotification_wp)+
+   apply (rule valid_tcbs_valid_tcbE, simp, simp)
+   apply clarsimp
+   apply (intro conjI)
+     apply (clarsimp simp: valid_tcb_def valid_bound_obj_def split: option.splits)
+    apply clarsimp
+   apply clarsimp
```
Comment (michaelmcinerney) [inline proof/refine/ARM/Ipc_R.thy:3132]: Can the `intro conjI` and the clarsimps be removed and replaced with just the one clarsimp?

**Review by michaelmcinerney — APPROVED** [review]
Nice

File: proof/refine/ARM/Ipc_R.thy [hunk 9]
```diff
@@ -2990,6 +2990,163 @@ lemma do_nbrecv_failed_transfer_corres:
...
+  done
+
+lemma set_sc_obj_resff_valid_tcb[wp]:
+  "thread_set f tptr \<lbrace>valid_tcb ptr tcb\<rbrace>"
+  apply (clarsimp simp: thread_set_def)
+  apply_trace (wpsimp wp: get_object_wp)
+  done
+
+lemma thread_set_valid_tcbs[wp]:
+ "\<lbrace>valid_tcbs and tcb_at tptr and
```
Comment (corlewis) [inline proof/refine/ARM/Ipc_R.thy:3031]: nitpick: indentation

File: proof/refine/ARM/Ipc_R.thy [hunk 10]
```diff
@@ -2990,6 +2990,163 @@ lemma do_nbrecv_failed_transfer_corres:
...
+  "\<lbrakk> (s, s') \<in> sr; cross_rel_ul sr P Q; P s \<longrightarrow> Q s' \<Longrightarrow> R \<rbrakk> \<Longrightarrow> R"
+  by (simp add: cross_rel_ul_def)
+
+lemma thread_set_weak_valid_sched_action2:
+  "\<lbrace>weak_valid_sched_action and scheduler_act_not tptr\<rbrace> thread_set f tptr \<lbrace>\<lambda>rv. weak_valid_sched_action\<rbrace>"
+  apply (wpsimp wp: thread_set_wp simp: obj_at_kh_kheap_simps vs_all_heap_simps fun_upd_def
+                    weak_valid_sched_action_def)
+  apply (auto simp: weak_valid_sched_action_def scheduler_act_not_def)
+  apply (rule_tac x=ref' in exI; simp)
+  apply clarsimp
```
[the remark this answers is not in the document]
Comment (corlewis) [reply proof/refine/ARM/Ipc_R.thy:3068]: Not just if possible, it's a firm rule that we don't want to maintain non-terminating autos.

**Review by corlewis — APPROVED** [review]
A bit of cleanup is needed, but otherwise looks good!

**Comment by xurtis** [comment]
If you want me to review this, could you give a high-level description in the PR description?

File: proof/refine/ARM/Ipc_R.thy (hunk 8 above)
Comment (MitchellBuckley) [reply proof/refine/ARM/Ipc_R.thy:3132]: Weirdly, no. A number of other variation seem not to work.

File: proof/refine/ARM/Ipc_R.thy [hunk 11]
```diff
@@ -3014,6 +3014,153 @@ lemma do_nbrecv_failed_transfer_corres:
...
+  "\<lbrace>weak_valid_sched_action and scheduler_act_not tptr\<rbrace> thread_set f tptr \<lbrace>\<lambda>rv. weak_valid_sched_action\<rbrace>"
+  apply (wpsimp wp: thread_set_wp simp: obj_at_kh_kheap_simps vs_all_heap_simps fun_upd_def
+                    weak_valid_sched_action_def)
+  apply (clarsimp simp: weak_valid_sched_action_def scheduler_act_not_def)
+  apply (rule_tac x=ref' in exI; clarsimp)
+  done
+
+find_theorems "ksReadyQueues" name: def
+
+definition notQueued :: "obj_ref \<Rightarrow> kernel_state \<Rightarrow> bool" where
```
Comment (mktnk3) [inline proof/refine/ARM/Ipc_R.thy:3085]: Is this a duplicate now?

**Review by mktnk3 — APPROVED** [review]
sorry I have a few late nitpicks. Apart from those, it looks great :)

File: proof/refine/ARM/Ipc_R.thy [hunk 12]
```diff
@@ -3014,6 +3014,153 @@ lemma do_nbrecv_failed_transfer_corres:
...
+
+lemma thread_set_weak_valid_sched_action2:
+  "\<lbrace>weak_valid_sched_action and scheduler_act_not tptr\<rbrace> thread_set f tptr \<lbrace>\<lambda>rv. weak_valid_sched_action\<rbrace>"
+  apply (wpsimp wp: thread_set_wp simp: obj_at_kh_kheap_simps vs_all_heap_simps fun_upd_def
+                    weak_valid_sched_action_def)
+  apply (clarsimp simp: weak_valid_sched_action_def scheduler_act_not_def)
+  apply (rule_tac x=ref' in exI; clarsimp)
+  done
+
+find_theorems "ksReadyQueues" name: def
```
Comment (mktnk3) [inline proof/refine/ARM/Ipc_R.thy:3083]: Remove

File: proof/refine/ARM/Ipc_R.thy (hunk 11 above)
Comment (MitchellBuckley) [reply proof/refine/ARM/Ipc_R.thy:3085]: Yes good point!

File: proof/refine/ARM/Ipc_R.thy (hunk 12 above)
Comment (MitchellBuckley) [reply proof/refine/ARM/Ipc_R.thy:3083]: Yes good point!

**Pull request merged**